ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/BlockchainQueries.java

```diff
@@ -165,6 +181,20 @@ public long headBlockNumber() {
     return fromAccount(address, blockNumber, Account::getBalance, Wei.ZERO);
   }
 
+  /**
+   * Returns the balance of the given account at a specific block hash.
+   *
+   * @param address The address of the account being queried.
+   * @param blockHash The block hash being queried.
+   * @return The balance of the account in Wei.
+   */
+  public Optional<Wei> accountBalance(final Address address, final Hash blockHash) {
+    final long blockNumber =
+        getBlockHeaderByHash(blockHash).map(BlockHeader::getNumber).orElse(Long.MAX_VALUE);
+
+    return accountBalance(address, blockNumber);
```

Comment:
Again, same concern with translating a block hash to a block number.  We return possibly false info for non canonical hashes.

`accountBalance` and the whole call series needs to be primarily hash based, where the calls to get it by block number quickly go to the hash based call chain.